service/worker/scanner/workflow/scavenger.go

```diff
@@ -0,0 +1,345 @@
+	//
+	// However, step 2 can be quite complicated:
+	//Here are all the cases that a workflow run(with unique runID) should be considered as "corrupted":
+	//a) a run has both current record and mutableState, workflowState is open, but history is corrupted
+	//b) a run has both current record and mutableState, workflowState is closed, but history is corrupted
+	//c) a run has only current record but no mutableState
+	//d) a run has only mutableState but no current record, workflow state is open
+	//e) a run with completed history without corruption, but timerMap/activityMap/etc are corrupted.
+	//
+	// Today this Scavenger only detecting for a) without performing actions on it.
+	// TODO https://github.com/uber/cadence/issues/2926 for more details.
+	Scavenger struct {
+		shardDB     p.ShardManager
+		historyDB   p.HistoryManager
+		resource    resource.Resource
+		hbd      ScavengerHeartbeatDetails
+		rps      int
+		numShards int
+		limiter  *rate.Limiter
+		metrics  metrics.Client
+		logger   log.Logger
+		isInTest bool
+	}
+
+	taskDetail struct {
+		shardID    int
+		domainID   string
+		workflowID string
+		runID      string
+
+		// passing along the current heartbeat details to make heartbeat within a task so that it won't timeout
+		hbd ScavengerHeartbeatDetails
+	}
+
+	taskResult struct{
+		err  error
+		corrupted bool
+	}
+
+)
+
+
+
+const (
+	// used this to decide how many goroutines to process
+	rpsPerConcurrency = 50
+	pageSize          = 1000
+	// clean up mutable state records if the started time is older than this threshold
+	cleanUpThreshold = time.Hour * 24 * 365
+)
+
+// NewScavenger returns an instance of history scavenger daemon
+// The Scavenger can be started by calling the Run() method on the
+// returned object. Calling the Run() method will result in one
+// complete iteration over all of the history branches in the system. For
+// each branch, the scavenger will attempt
+//  - describe the corresponding workflow execution
+//  - deletion of history itself, if there are no workflow execution
+func NewScavenger(
+	shardDB   p.ShardManager,
+	historyDB p.HistoryManager,
+	resource resource.Resource,
+	rps int,
+	numShards int,
+	hbd ScavengerHeartbeatDetails,
+	metricsClient metrics.Client,
+	logger log.Logger,
+) *Scavenger {
+
+	rateLimiter := rate.NewLimiter(rate.Limit(rps), rps)
+
+	return &Scavenger{
+		shardDB:shardDB,
+		historyDB:historyDB,
+		resource:resource,
+		hbd:     hbd,
+		rps:     rps,
+		numShards:numShards,
+		limiter: rateLimiter,
+		metrics: metricsClient,
+		logger:  logger,
+	}
+}
+
+var persistenceOperationRetryPolicy = common.CreatePersistanceRetryPolicy()
+
+// Run runs the scavenger
+func (s *Scavenger) Run(ctx context.Context) (ScavengerHeartbeatDetails, error) {
+	taskCh := make(chan taskDetail, pageSize)
+	respCh := make(chan taskResult, pageSize)
+	concurrency := s.rps/rpsPerConcurrency + 1
+
+	for i := 0; i < concurrency; i++ {
+		go s.startTaskProcessor(ctx, taskCh, respCh)
```

Comment:
consider using the Executor class in service/worker/executor package. If you use that package, you don't need to manually manage go routine pool / channels etc. However, the heartbeating mechanism used here needs to be modified a bit (which I think should be fine)